### PR TITLE
lldp: Power via MDI TLV: Initial support for 802.3bt

### PR DESCRIPTION
**Pull request by patrikdanielssonaxis**
Initial support for sending and receiving the new frames specified in 802.3bt. The type_ext field is used to differentiate 802.3bt, with zero value meaning no bt support.

Tested and verified in conjunction with a Cisco Catalyst 9300 switch.

### CODE REVIEW

File: src/lib/lldpctl.h [hunk 1]
```diff
@@ -708,8 +708,24 @@ typedef enum {
 	lldpctl_k_dot3_power_type, /**< `(I,W)` 802.3AT power type */
 	lldpctl_k_dot3_power_source, /**< `(IS,W)` 802.3AT power source */
 	lldpctl_k_dot3_power_priority, /**< `(IS,W)` 802.3AT power priority */
+    lldpctl_k_dot3_power_requested, /**< `(I,W)` 802.3AT power requested */
 	lldpctl_k_dot3_power_allocated, /**< `(I,W)` 802.3AT power allocated */
```
Comment (vincentbernat) [inline src/lib/lldpctl.h]: Do not reorder stuff, it makes the change backward-incompatible.

File: src/daemon/frame.h [hunk 2]
```diff
@@ -99,6 +99,11 @@ static union {
 #define PEEK_RESTORE POKE_RESTORE
 
 /* LLDP specific. We need a `tlv' pointer. */
+#define POKE_START_LLDP_TLV_8023BT(type)  \
+        (			   \
+	 tlv = pos,		   \
+	 POKE_UINT16((type << 9) | (0x1d))	   \
+	)
```
Comment (vincentbernat) [inline src/daemon/frame.h]: It's only used once, it doesn't seem worth it.

File: src/client/display.c [hunk 3]
```diff
@@ -469,6 +469,63 @@ display_port(struct writer *w, lldpctl_atom_t *port, int details)
...
 			}
 
+			/* 802.3bt */
+			if (lldpctl_atom_get_int(dot3_power,
+				lldpctl_k_dot3_power_type_ext) > LLDP_DOT3_POWER_8023BT_OFF) {
+				tag_start(w, "requested-a", "Requested mode a");
+				tag_data(w, lldpctl_atom_get_str(dot3_power,
+					lldpctl_k_dot3_power_requested_a));
+				tag_end(w);
+				tag_start(w, "requested-b", "Requested mode b");
```
Comment (vincentbernat) [inline src/client/display.c]: `"Requested mode B"`?

**Comment by vincentbernat** [comment]
It looks good. If you fancy it, you could update the integration test or I could do it later.

File: src/daemon/frame.h (hunk 2 above)
Comment (patrikdanielssonaxis) [reply src/daemon/frame.h]: Is your suggestion that the contents of the macro should be done in line in lldp.c instead?

Comment (vincentbernat) [reply src/daemon/frame.h]: Yes.

**Comment by vincentbernat** [comment]
Do you have a PCAP of an LLDPDU with 802.3BT TLVs?

Also, as for 802.3at, it seems IEEE doesn't provide the MIB, so we don't know the OID that should be used as a base. Do you happen to know if it exists somewhere?

**Comment by patrikdanielssonaxis** [comment]
??Sure I can record some frames captured on wireshark. Found one reference to MIB in the standard, and that was to the standard IEEE P802.3.1 (IEEE 802.3be), http://www.ieee802.org/3/be/, is that of any aid?


BR

Patrik Danielsson


[https://www.axis.com/mail_images/axis_logo_97x38px.gif]<https://www.axis.com/>

Axis Communications AB
Emdalavägen 14
SE-223 69 Lund
Sweden
Tel: +46 46 272 18 00
Fax: +46 46 13 61 30
www.axis.com<https://www.axis.com>

Patrik Danielsson
Software Engineer


e-mail: patrik.danielsson@axis.com<mailto:patrik.danielsson@axis.com>
Tel (direct): +46 46 272 12 36
Mobile: +46 70 497 33 86

[Axis Communications's Youtube Channel]<https://axis.com/blog>  [I'm on LinkedIn] <https://www.linkedin.com/company/6390>   [Axis Communications's Youtube Channel] <http://www.youtube.com/axiscommunications>   [Axis Communications's Facebook Fanpage] <http://www.facebook.com/axiscommunications>   [Follow Axis Communications on Twitter] <http://www.twitter.com/axisipvideo>   [Axis Communications's RSS Feed] <http://www.axis.com/rss>


P Please consider the environment before printing this e-mail.

*****Email Confidentiality Notice*****
This communication and any attached files are confidential and are only intended for the use of the individual or entity to which they are directed. They may contain information that is privileged and exempt from disclosure under applicable law. If an addressing or transmission error has misdirected this e-mail, please notify the sender and delete it from your system. If you are not the intended recipient you may not use, disclose, distribute, copy, print or rely on this e-mai

________________________________

**Comment by patrikdanielssonaxis** [comment]
Hi Vincent,

Here are some Wireshark captures of an LLDP negotiation.


BR

Patrik


[https://www.axis.com/mail_images/axis_logo_97x38px.gif]<https://www.axis.com/>

Axis Communications AB
Emdalavägen 14
SE-223 69 Lund
Sweden
Tel: +46 46 272 18 00
Fax: +46 46 13 61 30
www.axis.com<https://www.axis.com>

Patrik Danielsson
Software Engineer


e-mail: patrik.danielsson@axis.com<mailto:patrik.danielsson@axis.com>
Tel (direct): +46 46 272 12 36
Mobile: +46 70 497 33 86

[Axis Communications's Youtube Channel]<https://axis.com/blog>  [I'm on LinkedIn] <https://www.linkedin.com/company/6390>   [Axis Communications's Youtube Channel] <http://www.youtube.com/axiscommunications>   [Axis Communications's Facebook Fanpage] <http://www.facebook.com/axiscommunications>   [Follow Axis Communications on Twitter] <http://www.twitter.com/axisipvideo>   [Axis Communications's RSS Feed] <http://www.axis.com/rss>


P Please consider the environment before printing this e-mail.

*****Email Confidentiality Notice*****
This communication and any attached files are confidential and are only intended for the use of the individual or entity to which they are directed. They may contain information that is privileged and exempt from disclosure under applicable law. If an addressing or transmission error has misdirected this e-mail, please notify the sender and delete it from your system. If you are not the intended recipient you may not use, disclose, distribute, copy, print or rely on this e-mai

________________________________

**Comment by vincentbernat** [comment]
If you have attached it to the email, GitHub may have stripped it. Either attach them through GitHub interface or send them directly to me.

**Comment by stannous** [comment]
FYI, we tested these patches and do what we need.  Thanks for the patch!
Any idea when this will get pulled into master?

**Comment by vincentbernat** [comment]
I want to add tests (because if I only to add them later, I will never do it). Initially, I wanted also to add CLI support to set the values, but I did not set my mind over the keywords to use for that. Maybe @patrikdanielssonaxis could propose how to extend the `configure dot3 power` command? For the tests, I can just use the pcap file I've got. I'll do that over the week-end.

**Comment by patrikdanielssonaxis** [comment]
Sure, I can look into that!

On 19 Oct 2019, at 08:36, Vincent Bernat <notifications@github.com<mailto:notifications@github.com>> wrote:


I want to add tests (because if I only to add them later, I will never do it). Initially, I wanted also to add CLI support to set the values, but I did not set my mind over the keywords to use for that. Maybe @patrikdanielssonaxis<https://github.com/patrikdanielssonaxis> could propose how to extend the configure dot3 power command? For the tests, I can just use the pcap file I've got. I'll do that over the week-end.

**Comment by patrikdanielssonaxis** [comment]
Hi Vincent,
The solution that seems preferable is to do it in a similar way to 802.3at, i.e. to add another additional section for 802.3bt. Also to explain that the type_ext is incremented by 1 to have 0 mean no bt support. Would you like me to do it through a pull request or what is the best way to proceed?

BR
Patrik Danielsson

**Comment by vincentbernat** [comment]
Currently, the "signature" is:
```
       configure [ports ethX [,...]] dot3 power pse | pd [supported] [enabled] [paircontrol] powerpairs
       powerpairs [class class] [type type source source priority priority requested requested allocated
       allocated]
```

Just telling me how the updated signature should be is enough. Of course, if you want to go the extra mile and do a PR, you are welcome! I think that if the port is configured without additional arguments, it means BT is not supported. Would that be sensible and would you expect to advertise BT support from the existing AT values?

**Comment by patrikdanielssonaxis** [comment]
Hi Vincent,
I started out doing a PR but found it tricky with the .in files and figured it might be easier if you handled it if that is alright with you.

I am suggesting a change to the section in the manpage article to this:
```
configure [ports ethX [,...]] dot3 power pse | pd [supported] [enabled]
       [paircontrol] powerpairs powerpairs [class class] [type type source source priority
       priority requested requested allocated allocated] [pd_4pid pd_4pid requested_a requested_a
       requested_b requested_b allocated_a allocated_a allocated_b allocated_b
       pse_status pse_status pd_status pd_status pse_pairs_ext pse_pairs_ext class_a class_a
       class_b class_b class_ext class_ext power_type_ext power_type_ext pd_load pd_load
       pse_max pse_max]


Advertise Dot3 POE-MDI TLV for the given port or for all ports if none was
provided. One can act as a PD (power consumer) or a PSE (power provider). This
configuration is distinct of the configuration of the transmission of the
LLDP-MED POE-MDI TLV but the user should ensure the coherency of those two
configurations if they are used together.

supported means that MDI power is supported on the given port while enabled
means that MDI power is enabled.  paircontrol is used to indicate if pair
selection can be controlled. Valid values forr powerpairs are:
  signal  The signal pairs only are in use.
  spare   The spare pairs only are in use.

When specified, class is a number between 0 and 4.

The remaining parameters are optional and they consist of two sets. The first set
of parameters is for the 802.3at additions and the second one for the 802.3bt additions.
Please note that the 802.3at fields are required to be specified in order to use the
802.3bt fields.

type should be either 1 or 2, indicating which if the device conforms to
802.3at type 1 or 802.3at type 2. Values of source and priority are the same
as for LLDP-MED POE-MDI TLV.  requested and allocated are expressed in milliwatts.

pd_4pid is 1 or 0 signaling whether the PD supports powering of both modes.
allocated a and b and requested a and b are the specified values in milliwatts
for each pairset, this is only used for dual signature PDs. pse_status indicates
4 or 2 pair powering and dual or single signature on the PSE side while pd_status
indicates the same thing on the PD side. pse_pairs_ext shows alternative A, B or
both A and B power pairs. class a, b and extra signals power class requested
from the PD in dual signature mode for class_a and class_b and in single signature
mode for class_ext. 

Note that power_type_ext shows the PD type, and that this value is incremented by 1
to allow 0 to mean no 802.3bt support. pd_load is 1 or 0 and indicates whether the modes
are electrically isolated or not. pse_max is the maximum value for the PSE expressed
in milliwatts.

Here are three valid uses of this command:
      configure ports eth3 dot3 power pse supported enabled paircontrol
      powerpairs spare class class-3

      configure dot3 power pd supported enabled powerpairs spare class class-3
      type 1 source pse priority low requested 10000 allocated 15000

      configure dot3 power pd type 2 source pse priority low requested 71000
      allocated 0 pd_4pid 1 requested_a 35500 requested_b 35500 
      allocated_a 0 allocated_b 0 pd_status 3 class_a 5 class_b 5 class_ext 15
      power_type_ext 6 pd_load 0
```
802.3bt support is advertised in the code through the "power_type_ext" field, with the value expected being incremented by 1 compared to the standard, in a similar manner to how power_type is used for 802.3at.

I also have a question regarding the required parameters of the configure command. powerpairs is required as an argument but its value when transmitted from a PD is undefined, so maybe it should be optional?

Best Regards
Patrik Danielsson

**Comment by vincentbernat** [comment]
❦ 21 octobre 2019 05:59 -07, patrikdanielssonaxis <notifications@github.com>:

> I also have a question regarding the required parameters of the
> configure command. powerpairs is required as an argument but its value
> when transmitted from a PD is undefined, so maybe it should be
> optional?

I don't remember why it was marked as mandatory. Let's change that.

**Comment by vincentbernat** [comment]
❦ 21 octobre 2019 05:59 -07, patrikdanielssonaxis <notifications@github.com>:

> configure [ports ethX [,...]] dot3 power pse | pd [supported] [enabled]
>        [paircontrol] powerpairs powerpairs [class class] [type type source source priority
>        priority requested requested allocated allocated] [pd_4pid pd_4pid requested_a requested_a
>        requested_b requested_b allocated_a allocated_a allocated_b allocated_b
>        pse_status pse_status pd_status pd_status pse_pairs_ext pse_pairs_ext class_a class_a
>        class_b class_b class_ext class_ext power_type_ext power_type_ext pd_load pd_load
>        pse_max pse_max]

OK, looks good. I'll replace underscores by dashes.

Do we need to implement more constraints on values? Does 802.3AT stuff
should be here when declaring 802.3BT or is that not mandatory? Also,
should everything be there or is everything optional or is there
something between that? For example, is pd_load only for PD?

**Comment by patrikdanielssonaxis** [comment]
Sure! Yes if declaring 802.3bt fields then 802.3at is required as bt is just an extension to at. Well it is tricky with the optional values for 802.3at and 802.3bt as which values are required differs depending on if it's a PD or a PSE. It might be preferable to just keep it as is, that is, keep the entire set optional, and if a value is not specified in a set then it should be considered to have a value of 0.

**Comment by vincentbernat** [comment]
For 802.3AT, I have pushed a modification to only allow powerpairs when PSE. We can do the same for 802.3BT if you tell me which field is allowed in which conditions.

**Comment by patrikdanielssonaxis** [comment]
Sure!
Optional fields:
requested-a, requested-b, allocated-a, allocated-b, pd-load

Required when PSE:
pse-status, pse-pairs-ext, pse-max

Required when PD:
pd-status

Always required:
class-a, class-b, class-ext, power-type-ext

**Comment by patrikdanielssonaxis** [comment]
Forgot that also pd-4pid is required when PD.

**Comment by sudhank29** [comment]
Thanks very much for this patch!
Is there any update on extending the `configure dot3 power` command ?

**Comment by vincentbernat** [comment]
Didn't get time yet. But I intend to do that before releasing a new version.

**Comment by sudhank29** [comment]
OK thanks for the update. Any approximate timeline for the new version?

**Comment by bschaef14** [comment]
Hello @vincentbernat 

Just saw that a new release occured. Glad 802.3BT support got included! I think lldpcli support for new power TLVs might've fallen off the radar though.

**Comment by vincentbernat** [comment]
It's still on my todo list, but I didn't get time to do it. I didn't want to hold the next release for this.

**Comment by andreas-hofmann** [comment]
Hi @vincentbernat, I have just created pull request #449 with a proposal for 802.3bt parameter support in lldpcli. I took the liberty to use the proposed naming schemes in this ticket as guidelines. If you have time to review it, any feedback would be highly appreciated!

**Pull request merged**